core/src/common.cpp

```diff
@@ -362,6 +364,434 @@ namespace irods::http
+
+		auto json_res{hit_introspection_endpoint(url_encode_body(args))};
+
+		// Validate access token
+		if (!json_res.at("active").get<bool>()) {
+			logging::warn("{}: Access token is invalid or expired.", __func__);
+			return std::nullopt;
+		}
+
+		return json_res;
+	}
+
+	/// Fetches JWKs from the location specified by the OpenID Provider
+	///
+	/// See OpenID Connect Discovery 1.0 Section 3 for info on jwks_uri
+	/// See RFC 7517 for more information on JSON Web Key (JWK)
+	///
+	/// \returns A std::string represting the JWKs from the OpenID Provider
+	auto fetch_jwks_from_openid_provider() -> std::string
+	{
+		namespace logging = irods::http::log;
+
+		const auto jwks_uri{
+			irods::http::globals::oidc_endpoint_configuration().at("jwks_uri").get_ref<const std::string&>()};
+
+		const auto parsed_uri{boost::urls::parse_uri(jwks_uri)};
+
+		if (parsed_uri.has_error()) {
+			logging::error("{}: Error trying to parse jwks_uri [{}]. Please check configuration.", __func__, jwks_uri);
+			throw std::runtime_error{"Invalid [jwks_uri]."};
+		}
+
+		const auto url{*parsed_uri};
+		const auto port{get_port_from_url(url)};
+
+		// Addr
+		net::io_context io_ctx;
+		auto tcp_stream{irods::http::transport_factory(url.scheme_id(), io_ctx)};
+		tcp_stream->connect(url.host(), *port);
+
+		// Build Request
+		constexpr auto http_version_number{11};
+		beast::http::request<beast::http::string_body> req{beast::http::verb::get, url.path(), http_version_number};
+		req.set(beast::http::field::host, irods::http::create_host_field(url, *port));
+		req.set(beast::http::field::user_agent, irods::http::version::server_name);
+		req.set(beast::http::field::accept, "application/json");
+		req.prepare_payload();
+
+		// Send request & receive response
+		auto res{tcp_stream->communicate(req)};
+
+		logging::debug("{}: Received the following response: [{}]", __func__, res.body());
+
+		return res.body();
+	}
+
+	/// Adds the specified symmetric algorithm \p _alg to the verifier \p _verifier, using the secrets provided in the
+	/// HTTP API configuration
+	///
+	/// See RFC 7518 for details on JSON Web Algorithms (JWA)
+	///
+	/// \param[in,out] _verifier The jwt::verifier to add additional verification algorithms to.
+	/// \param[in]     _alg      The signing algorithm requested by the signed JWT.
+	auto add_symmetric_alg(
+		jwt::verifier<jwt::default_clock, jwt::traits::nlohmann_json>& _verifier,
+		std::string_view _alg) -> void
+	{
+		namespace logging = irods::http::log;
+
+		auto algorithm_family{_alg.substr(0, 2)};
+		auto algorithm_specifics{_alg.substr(2)};
+
+		// Extract the secret key
+		// OpenID secret should be our secret
+		std::string key;
+		if (auto realm_secret{irods::http::globals::oidc_configuration().find("realm_secret")};
+		    realm_secret != std::end(irods::http::globals::oidc_configuration()))
+		{
+			key = jwt::base::decode<jwt::alphabet::base64url>(jwt::base::pad<jwt::alphabet::base64url>(realm_secret->get<std::string>()));
+		}
```

Comment:
Should we have `realm_secret` be a base64url or base64 encoded string?

Defaulted to base64url as that's what keycloak uses. The key should be encoded, as not all keys are ASCII printable.

Reply:
> Should we have `realm_secret` be a base64url or base64 encoded string?

Is base64 encoding it a requirement?
What are the pros/cons of each encoding?

What is the `realm_secret` for?

> Defaulted to base64url as that's what keycloak uses. The key should be encoded, as not all keys are ASCII printable.

Is the decision to base64 encode the key strictly for printing it?

Reply:
being able to print it helps with debugging... 

are there other use cases?

Reply:
Why we need an encoding of some sort:

``` python
>>> # This a base64url encoded key from keycloak
>>> base64.urlsafe_b64decode('ACu4YsAjVil7KC3G9dVL_THazeFa4AVDZ83lNPiKV4KxEXc9tzrfeUzGJVvGsNG2oWuSLq18tIo4i8VUH6Dkpw==')
b'\x00+\xb8b\xc0#V){(-\xc6\xf5\xd5K\xfd1\xda\xcd\xe1Z\xe0\x05Cg\xcd\xe54\xf8\x8aW\x82\xb1\x11w=\xb7:\xdfyL\xc6%[\xc6\xb0\xd1\xb6\xa1k\x92.\xad|\xb4\x8a8\x8b\xc5T\x1f\xa0\xe4\xa7'
>>> # The key is not 'ASCII Printable'.
>>> base64.urlsafe_b64decode('ACu4YsAjVil7KC3G9dVL_THazeFa4AVDZ83lNPiKV4KxEXc9tzrfeUzGJVvGsNG2oWuSLq18tIo4i8VUH6Dkpw==').decode('ascii')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xb8 in position 2: ordinal not in range(128)
```

`realm_secret` needs to be given to us through the config. I'm not sure the JSON parser can deal with the gibberish output. Additionally, it may be prone to errors if the key were to change at some point.

Reply:
Got it. Carry on.

Reply:
Alright, we'll leave as a base64url encoded string, and leave a comment on the readme.

Reply:
Yep, please resolve once the comment is added to the README.

Reply:
Done.